**Universidad Internacional de La Rioja (UNIR) - Máster Universitario en Inteligencia Artificial - Procesamiento del Lenguaje Natural**

***
Datos del alumno (Nombre y Apellidos): Diego Alexander Torres Forero

Fecha:20/04/2026
***

<span style="font-size: 20pt; font-weight: bold; color: #0098cd;">Trabajo: Caracterización de textos</span>

**Objetivos**

Con esta actividad se tratará de que el alumno se familiarice con el manejo de la librería spacy, así como con los conceptos básicos de manejo de las técnicas NER

**Descripción**

En esta actividad debes procesar de forma automática un texto en lenguaje natural para detectar características básicas en el mismo, y para identificar y etiquetar las ocurrencias de conceptos como localización, moneda, empresas, etc.

En la primera parte del ejercicio se proporciona un código fuente a través del cual se lee un archivo de texto y se realiza un preprocesado del mismo. En esta parte el alumno tan sólo debe ejecutar y entender el código proporcionado.

En la segunda parte del ejercicio se plantean una serie de preguntas que deben ser respondidas por el alumno. Cada pregunta deberá responderse con un fragmento de código fuente que esté acompañado de la explicación correspondiente. Para elaborar el código solicitado, el alumno deberá visitar la documentación de la librería spacy, cuyos enlaces se proporcionarán donde corresponda.

# Parte 1: carga y preprocesamiento del texto a analizar

Observa las diferentes librerías que se están importando.

In [ ]:
import pathlib
import spacy
import pandas as pd
from spacy import displacy
import csv
import es_core_news_md

El siguiente código simplemente carga y preprocesa el texto. Para ello, lo primero que hace es cargar un modelo de lenguaje previamente entrenado. En este caso, se utiliza <i>es_core_news_md</i>:

https://spacy.io/models/es#es_core_news_md


In [ ]:
nlp = es_core_news_md.load()

El objeto <i>nlp</i> permite utilizar el modelo de lenguaje cargado, de forma que se puede procesar un texto y obtenerlo en su versión preprocesada. Así, nos permite realizar las diferentes tareas. En este caso, vamos a utilizar el pipeline para hacer un preprocesamiento básico, que consiste en tokenizar el texto.

In [10]:
from google.colab import files
uploaded = files.upload()

filename = "./02Dataset_anonimizado.csv"
lines_number = 10000

data = pd.read_csv(
    filename,
    sep = ";",
    nrows = lines_number,
    encoding = "latin-1"
)

data["odio_binario"] = data["INTENSIDAD"].apply(lambda x: 0 if x == 0 else 1)

data = data.loc[:, ~data.columns.str.contains('^Unnamed')]

print(data.head())
print(data.columns)
print(data.shape)

      MEDIO SOPORTE           URL TIPO DE MENSAJE  \
0  EL PAÍS      WEB  URL_a4d7efc0      COMENTARIO   
1  EL PAÍS      WEB  URL_a4d7efc0      COMENTARIO   
2  EL PAÍS      WEB  URL_54312d9e      COMENTARIO   
3  EL PAÍS      WEB  URL_54312d9e      COMENTARIO   
4  EL PAÍS      WEB  URL_54312d9e      COMENTARIO   

                                CONTENIDO A ANALIZAR  INTENSIDAD TIPO DE ODIO  \
0  el barí§a nunca acaeza ante un segundo b ni an...         3.0        Otros   
1  el real madrid ha puesto punto y final a su an...         0.0          NaN   
2  cristina cifuentes podrí­a haber sido la presi...         3.0   Ideológico   
3  habrí­a que reabrir el caso. el supremo se ded...         3.0   Ideológico   
4  me parece un poco exagerado pedir más de tres ...         3.0   Ideológico   

  TONO HUMORISTICO MODIFICADOR  odio_binario  
0              NaN         NaN             1  
1              NaN         NaN             0  
2              NaN         NaN             1  
3     

El código anterior carga el archivo CSV (opcionalmente con un límite de líneas a leer) y genera la variable <i>data</i>, que contiene un Dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html) con los datos leídos del CSV.

Te vendrá bien conocer la siguiente documentación:
<ul>
    <li>https://spacy.io/api/doc</li>
    <li>https://spacy.io/api/token</li>
    <li>https://spacy.io/api/morphology#morphanalysis</li>
</ul>

### Playground

Utiliza este espacio para hacer pruebas y ensayos con las variables generadas con el código previo. A modo de ejemplo, se ofrece código que realiza las siguientes tareas:


- leer un número dado de líneas del Dataframe y generar dos listas con los valores (se pueden leer directamente del DataFrame, se muestra el ejemplo como una opción más)
- procesar el texto de cada comentario


Para procesarlo, hay utilizar el objeto <i>nlp</i> y así obtener objetos de la clase <i>Doc</i> (https://spacy.io/api/doc)

Visita la documentación de dicha clase y experimenta probando las diferentes funciones y atributos

In [ ]:
# Puedes insertar aquí código de pruebas para experimentar con las diferentes funciones y atributos de 'doc'.
#print(data["CONTENIDO A ANALIZAR"][1])
#print(data["INTENSIDAD"][1])
doc = []
value = []

#con el bucle, generamos sendas listas con los comentarios ya parseados y con el valor de intensidad
for i in range(0, lines_number):

    #en un primer paso se parsea el comentario. En el segundo paso se añade el objeto a la lista
    tmp_doc = nlp(data["CONTENIDO A ANALIZAR"][i])
    doc.append(tmp_doc)

    #en un primer paso extrae el valor. En el segundo paso se añade el valor a la lista
    tmp_value = data["INTENSIDAD"][i]
    value.append(tmp_value)


#ejemplo de cómo recorrer un comentario palabra por palabra
for token in doc[1]:
    print(token)

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 1.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántos registros contiene el corpus?</span>

In [12]:
recordsNumber = data.shape[0]
print("Numero registros:", recordsNumber)


Numero registros: 10000


<b>Para llegar a esta respuesta consulte el foro de stack ovetflow en el cual especifican que para pandas se usa el .shape[0] para obtener la cantidad de registros del dataset. Por lo que el limite impuesto al momento de cargar la data funcionó y cargo solamente 10.000 registros

Este es el link del foro del cuál me base para llegar a esta respuesta: https://stackoverflow.com/questions/17468878/pandas-python-how-to-count-the-number-of-records-or-rows-in-a-dataframe</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 2.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuántas palabras totales hay en los comentarios del corpus?</span>

In [13]:
def contarPalabras(text):
    doc = nlp(str(text))
    return sum(1 for token in doc if not token.is_space and not token.is_punct)

data["num_palabras"] = data["CONTENIDO A ANALIZAR"].apply(contarPalabras)

total_palabras = data["num_palabras"].sum()
print("Número total de palabras:", total_palabras)


Número total de palabras: 216496


<b>Para llegar a esta repsuesta me base en la documentación de spacy donde especifica que para poder procesar el texto se tokeniza es decir dividir el texto por palabras sin tener en cuenta los signos de puntuación, especios en blanco, entre otros.

Basta con conocer el numero de tokens por celda para conocer el numero de palabras de un solo registro y el resultado fué puesto en una nueva columna llamada num_palabras, para encontrar el total cómo lo solicita este punto tuve que sumar y generar un total basado en lso resultados puesttos en la nueva columna creada, este analisis se hizo sobre la columna [CONTENIDO A ANALIZAR] y el resultado obtenido fué de 216496 palabras.

Documentación de Spacy usada:

Tokenización https://spacy.io/usage/spacy-101

Características Lingüísticas: https://spacy.io/usage/linguistic-features
</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 3.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Cuál el número promedio de palabras en cada comentario?</span>

In [14]:
promedio_palabras = data["num_palabras"].mean()
print("Promedio de palabras por comentario:", promedio_palabras)

Promedio de palabras por comentario: 21.6496


<b>Para llegar a esta respuesta aproveche el ejercicio anterior para usar la nueva columna que contiene el numero de palabras por registro de esta manera solo tuve que calcular el promedio sobre la columna [num_palabras] al ejecutar esta parte del código se obtuvo un resultado de 21.65 palabras en promedio.


Para poder realizar la implementación me base en el foro de StackOverflow sobre cómo hallar promedio de una columna en un dataset con Pandas.

Recusro consultado: https://es.stackoverflow.com/questions/386319/pandas-promedio-columna-dataframe </b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 4.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál el número promedio de palabras en los comentarios de cada grupo?</span>

In [16]:
promedioPalabras = data.groupby("odio_binario")["num_palabras"].mean()
print(promedioPalabras)

print("Promedio No odio:", promedioPalabras[0])
print("Promedio Odio:", promedioPalabras[1])

odio_binario
0    21.548530
1    22.899733
Name: num_palabras, dtype: float64
Promedio palabras - No odio: 21.548530047557286
Promedio palabras - Odio: 22.899732620320854


<b>
El cargar el dataset puse la siguiente linea:

data["odio_binario"] = data["INTENSIDAD"].apply(lambda x: 0 if x == 0 else 1)

Con esto establecí una variable binaria de odio ya que el ejercicio exige saber si tiene o no tiene odio.

Para este punto en particular fué un poco confuso peor al final pude interpretar el concepto y calcular el promedio de palabras con odio y no odio de los registros con esto según la variable creada al cargar el dataset odio_binario pude calcular el promedio con ayuda de la columna num_palabras en cada grupo y asi pude encontrar si los mensajes con odio tienden a ser mas largos o más cortos que los mensajes sin odio.

para este caso tenemos un total de 21.55 palabras de no odio y 22.90 de palabras con odio por lo que según este resultado tienen a ser mas largos los mensajes con odio, lo que puede sugerir que los mensajes de odio pueden redactarse de una forma extensa en conparacion a los mensajes de no odio.

Para esta respuesta usé la documentación de Spacy de Sentencizer:

https://spacy.io/api/sentencizer


</b>

<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 5.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el número promedio de oraciones en los comentarios de cada grupo?</span>

In [19]:
def ContOraciones(texto):
    doc = nlp(str(texto))
    return sum(1 for sent in doc.sents)

data["NUM_ORACIONES"] = data["CONTENIDO A ANALIZAR"].apply(ContOraciones)

promedioOraciones = data.groupby("odio_binario")["NUM_ORACIONES"].mean()
print(promedioOraciones)


odio_binario
0    1.779615
1    1.770053
Name: NUM_ORACIONES, dtype: float64


<b>
Para responder a esta pregunta fué necesario el uso de spacy con al funcionalidad para segmentar oraciones mediante doc.sents. De esta manera se conto las oraciones que contiene cada comentario y este resultado se guardo en la columna NUM_ORACIONE. Luego procedí a realizar el calculo de oraciones para cada grupo.

De esta manera el resultado obtenido fué el siguiente:

Promedio de oraciones en los comentarios de no odio 1.78
Promedio de oraciones en los comentarios de no odio 1.77

El proposito de este ejercicio podria ser validar si los mensajes de odio podrían estar redactados como expresiones cortas  o cómo mensajes más elaborados.


Para este punto volvi a usar la documentacion de lisguistic features de spacy:

https://spacy.io/usage/linguistic-features#sbd
</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Pregunta 6.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Considerando dos grupos de comentarios (odio y no odio) ¿Cuál es el porcentaje de comentarios que contienen entidades NER en cada grupo?</span>

In [20]:
def contiene_entidades(texto):
    doc = nlp(str(texto))
    return 1 if len(doc.ents) > 0 else 0

data["tiene_entidades"] = data["CONTENIDO A ANALIZAR"].apply(contiene_entidades)

porcentaje_entidades = data.groupby("odio_binario")["tiene_entidades"].mean() * 100
print(porcentaje_entidades)


odio_binario
0    33.906182
1    37.566845
Name: tiene_entidades, dtype: float64


<b>

Para validar el porcentaje de comentarios que tiene entidades NER, lo que hice fué nuevamente procesar cada texto con spacy y valide si tiene al menos una entidad conocida por dov.ents Esta indormación la represente de forma binaria en una nueva columna llamada tiene_entidades donde el 1 es para presencia y el 0 para ausencia. Con este resultado luego calcule el porcentaje de cada grupo con ayuda del calculo del promedio multiplicado por 100 y con esto obtuve los siguientes resultados:

Ausencia de la entidad: 33.91%
presencia de la entidad: 37.57%

Para solucionar este ejercicio me base en la documentación de Spacy.

uso del doc.ents: https://spacy.io/api/doc#ents
reconocimientod e entidades: https://spacy.io/usage/linguistic-features#named-entities

</b>
<hr>


<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Plantea tus propias preguntas</span>

<span><b>Plantea al menos 4 características</b> del texto cuyo análisis permita una caracterización completa del texto. Puedes utilizar recomendaciones proporcionadas por la IA Generativa, si así lo deseas. Para cada una de las características planteadas, obtén valores separados para los grupos ODIO/NO-ODIO.</span>

<span>En la explicación aportada, deberás <b>explicar el significado de la característica planteada</b> así como la importancia de ésta en la caracterización del texto.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 1.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[Proporciond e mayúsculas]</span>

In [21]:
def proporcion_mayusculas(texto):
    palabras = str(texto).split()
    if len(palabras) == 0:
        return 0
    mayusculas = sum(1 for p in palabras if p.isupper() and len(p) > 1)
    return mayusculas / len(palabras)

data["prop_mayusculas"] = data["CONTENIDO A ANALIZAR"].apply(proporcion_mayusculas)

print(data["prop_mayusculas"].describe())
print(data.groupby("odio_binario")["prop_mayusculas"].mean())


count    10000.000000
mean         0.022019
std          0.086468
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: prop_mayusculas, dtype: float64
odio_binario
0    0.023122
1    0.008366
Name: prop_mayusculas, dtype: float64


<b>Las mayúsculas dentro de la redaccion digital se entiende como el hacer un enfasis, agresividad, contundencia, etc. Por esta razon considero que es necesaria para analizar este tipo de mensajes de odio y no odio

Para esto lo que hice fué contar las palabras que están en mayusculas y dividir ese resultado por la cantidad total de palabras de cada comentario.

Con el resultado obtenido pude determinar que no es una varable significativa para el estudio puesto que los valores fueron minimos por lo que su presencia para sacar conclusiones acertadas puede ser descartada.
</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 2.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[Cantidad de signos de exclamación o interrogación]</span>

In [22]:
def contar_signos_expresivos(texto):
    texto = str(texto)
    return texto.count("!") + texto.count("¡") + texto.count("?") + texto.count("¿")

data["signos_expresivos"] = data["CONTENIDO A ANALIZAR"].apply(contar_signos_expresivos)

print(data["signos_expresivos"].describe())
print(data.groupby("odio_binario")["signos_expresivos"].mean())

count    10000.000000
mean         0.613800
std          1.494012
min          0.000000
25%          0.000000
50%          0.000000
75%          1.000000
max         50.000000
Name: signos_expresivos, dtype: float64
odio_binario
0    0.608733
1    0.676471
Name: signos_expresivos, dtype: float64


<b>Para este ejercicio fué un poco mas sencillo ya que solo tuve que contar especificamente los simbolos de exclamación e interrogación.

Consideré esta metrica porque en el lenguaje escrito se usan para exaltar emociones, ironía, etc.

de esta manera encontre un promedio interesante de 1.50 lo que podría justificar un estudio mas profundo de esta métrica

</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 3.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[Longitud de las palabras]</span>

In [23]:
def longitud_media_palabras(texto):
    doc = nlp(str(texto))
    palabras = [token.text for token in doc if not token.is_space and not token.is_punct]
    if len(palabras) == 0:
        return 0
    return sum(len(p) for p in palabras) / len(palabras)

data["long_media_palabra"] = data["CONTENIDO A ANALIZAR"].apply(longitud_media_palabras)

print(data["long_media_palabra"].describe())
print(data.groupby("odio_binario")["long_media_palabra"].mean())


count    10000.000000
mean         5.059357
std          1.829656
min          0.000000
25%          4.200000
50%          4.692308
75%          5.377928
max         48.000000
Name: long_media_palabra, dtype: float64
odio_binario
0    5.063348
1    5.009999
Name: long_media_palabra, dtype: float64


<b>

La longitud puede dar indicios del tipo de lenguaje usado en los comentarios. Para esto se utilizo nuevamente la tokenicación que ofrece la libreria de spacy, nuevamente se eliminaron los signos de puntuación, espacios, etc. Se procedió a calcular el numero de caracteres de cada palabra. Lo que espero es que esto ayude a diferenciar mensajes sencillos a mensajes con mayor nivel de redacción.



</b>
<hr>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Característica adicional 4.</span>


<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">[Diversidad de palabras]</span>

In [24]:
def diversidad_lexica(texto):
    doc = nlp(str(texto))
    palabras = [token.text.lower() for token in doc if not token.is_space and not token.is_punct]
    if len(palabras) == 0:
        return 0
    return len(set(palabras)) / len(palabras)

data["diversidad_lexica"] = data["CONTENIDO A ANALIZAR"].apply(diversidad_lexica)

print(data["diversidad_lexica"].describe())
print(data.groupby("odio_binario")["diversidad_lexica"].mean())


count    10000.000000
mean         0.928250
std          0.096363
min          0.000000
25%          0.875000
50%          1.000000
75%          1.000000
max          1.000000
Name: diversidad_lexica, dtype: float64
odio_binario
0    0.928024
1    0.931045
Name: diversidad_lexica, dtype: float64


<b>

Propongo esta característica para determinar si los comentarios analizados contienen variedad en palabras o por el contrario son comentarios más bien uniformes repetitivos ya que en los mensajes de odio en particular las expresiones tienden a repetirse.

Para esto conte las palabrás únicas entre el número total de palabras esto después de normalizar el texto.

Con los resultados obtenidos podemos estar ante una variable que nos puede dar indicios para profundizar en el estudio de la variedad léxica usada en este tipo de mensajes.

</b>
<hr>


<span style="font-size: 16pt; font-weight: bold; color: #0098cd;">Reflexión final</span>

<span>Una de las utilidades de la caracterización de texto es que nos sirve como etapa de <i>feature-extraction</i> (extración de características) de cara a un posterior sistema de clasificación. Es pertinente, por tanto, reflexionar sobre la capacidad discriminatoria de cada una de las características extraídas. </span>

<span> Responde, para ello, a la siguiente pregunta.</span>

<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">Reflexión final.</span>
<span style="font-size: 14pt; font-weight: bold; color: #0098cd;">¿Es posible utilizar alguna de las características extraídas en las preguntas anteriores para determinar si un mensaje contiene odio? Justifica tu respuesta con el análisis estadístico que consideres necesario.</span>

In [26]:
caracteristicas = [
    "num_palabras",
    "NUM_ORACIONES",
    "tiene_entidades",
    "prop_mayusculas",
    "signos_expresivos",
    "long_media_palabra",
    "diversidad_lexica"
]

resumen = data.groupby("odio_binario")[caracteristicas].mean()
print(resumen)

# Correlación simple con la variable objetivo
correlaciones = data[caracteristicas + ["odio_binario"]].corr()["odio_binario"].sort_values(ascending=False)
print(correlaciones)

              num_palabras  NUM_ORACIONES  tiene_entidades  prop_mayusculas  \
odio_binario                                                                  
0                21.548530       1.779615         0.339062         0.023122   
1                22.899733       1.770053         0.375668         0.008366   

              signos_expresivos  long_media_palabra  diversidad_lexica  
odio_binario                                                            
0                      0.608733            5.063348           0.928024  
1                      0.676471            5.009999           0.931045  
odio_binario          1.000000
tiene_entidades       0.020303
signos_expresivos     0.011928
diversidad_lexica     0.008246
num_palabras          0.004982
NUM_ORACIONES        -0.001086
long_media_palabra   -0.007671
prop_mayusculas      -0.044898
Name: odio_binario, dtype: float64


<b>
Me incline por un analisis descriptivo mas una correlación.
Resultado:

tiene_entidades = 0.0203
signos_expresivos = 0.0119
diversidad_lexica = 0.0082
num_palabras = 0.0050
NUM_ORACIONES = -0.0011
long_media_palabra = -0.0077
prop_mayusculas = -0.0449


Con el anterior resultado es facil ver que por si sola una variable no es capaz de diferenciar entre odio y no odio, son muy cercanas entre si o muy cercanas al cero, por lo que no serían un referente válido y las correlaciones arrojadas no tienen la suficiente confiabilidad para tomar una conclusión sobre si tiene o no odio un mensaje.

La característica que presenta la relación más marcada, aunque todavía débil, es la proporción de mayúsculas (prop_mayusculas), con una correlación de -0.0449. Esto sugiere que, en este corpus, una mayor proporción de mayúsculas no se asocia con una mayor presencia de odio, sino más bien con lo contrario, aunque la intensidad de esta relación es demasiado baja como para considerarla determinante.

para el calculo y selección de estos metodos estadisticos, me base en la teoría de coeficiente de correlación que puede ser consultada en el siguente link:

https://www.statisticshowto.com/probability-and-statistics/correlation-coefficient-formula/



</b>
<hr>
